added two new methods clear_key_before and clear_key_after

niphlod opened the pull request:
Fixes issue #4

Those are meant mainly for easier maintenance purposes

Notes:
- new methods are meant mainly for easier maintenance purposes
- schema change needs a new version, plus migration scripts
- remember to add in the docs that using the new methods statistics are
  not updated

Unrelated: 
- switched to faster travis-ci infrastructure based on docker
- fails on newer pypy because pythonapi is no longer available

niphlod committed:
added two new methods clear_key_before and clear_key_after

Fixes issue #4

Those are meant mainly for easier maintenance purposes

Notes:
- new methods are meant mainly for easier maintenance purposes
- schema change needs a new version, plus migration scripts
- remember to add in the docs that using the new methods statistics are
  not updated

Unrelated: switched to faster travis-ci infrastructure based on docker

chriso commented on test/test_gauged.py:
```diff
@@ -549,6 +549,78 @@ def test_clear_from(self):
+        self.assertEqual(gauged.value('foo', timestamp=30000), 3)
+        self.assertEqual(gauged.value('bar', timestamp=30000), 3)
+        # 'foo' value before 20000 is cleared
+        self.assertEqual(gauged.value('foo', timestamp=10000), None)
```
Can you add a test for `gauged.value('foo', timestamp=20000)` here to show if timestamps are inclusive?

chriso replied:
Actually I see it tested below, nm.

chriso commented on test/test_gauged.py:
```diff
@@ -549,6 +549,78 @@ def test_clear_from(self):
+        # every value before 20000 stays the same
+        self.assertEqual(gauged.value('foo', timestamp=10000), 1)
+        self.assertEqual(gauged.value('bar', timestamp=10000), 1)
+        # 'foo' value after 20000 is cleared
```
Same here. Include a test for `gauged.value('bar', timestamp=20000)`.

chriso commented:
Looks good after an initial pass. The migration will need to clear out the cache before adding the column.

Thanks for switching over Travis to their container interface. I'm happy for you to drop pypy for now so that tests pass.

niphlod committed:
added tests to show inclusion, allow pypy build to fail

niphlod committed:
travis.yml fix

niphlod commented:
should be finenow . I avoided to trash completely testing on pypy because forgetting it would be a shame :P 
Should I include in this also migration-related adaptations or make a new PR ?

chriso commented:
Thanks, that's fine for pypy. I'll have to think of a better fix later or possibly even rip it out.

The migration should be part of this PR.

niphlod committed:
arrange imports at the top to make pypy working

niphlod committed:
initial plan to support migration

niphlod commented:
aside from test failing for mysql and postgresql for version mismatch, could you please take a look at the current status ?
Work done:
- fix imports for pypy
- added a new exception to signal a migration error
- added a common method to support migrations in drivers/interface.py, as I feel duplicating the logic in all backends can be avoided
- added a debug parameter to inspect what is really going on under the hoods in the migration
- sqlite migrate logic added. All that it takes is to have a script to upgrade the previous version to the new one. Unfortunately SQLite can't accept a NOT NULL column without a default value (even if empty), so the entire table needs to be dropped and recreated

chriso commented on .travis.yml:
```diff
@@ -13,3 +14,6 @@ before_script:
 script: make check
+matrix:
+  allow_failures:
+    - "python": "pypy"
```
Can you drop this now?

chriso commented:
This is looking good. I think a better approach would be to take all of the logic in your `driver.migrate(...)` function and move it to a helper script. This has two benefits, a) you're not putting `print` statements in library code and can instead use the Python logging interface with appropriate verbosity settings, and b) you're reusing as much code as possible across drivers.

```sh
$ gauged_migrate <config>
```

You'd define the migrations as you've done in `prepare_migrations` and then expose a way for the driver to run one or more migrations (all wrapped in a transaction).

niphlod commented:
got pretty much all the points, albeit the latest requiring multiple migrations in a single transaction wouldn't be possible. 
Any DDL statement can't be rollbacked safely, so I'd still go for executing every step in a single transaction.

niphlod committed:
supporting migrations, plan B

niphlod commented:
@chriso : please see the latest version and if you're happy with it.

chriso commented:
Looking good :+1:

niphlod committed:
added migrations for remaining drivers

niphlod commented:
ok, this should be the latest .

chriso commented on gauged/drivers/mysql.py:
```diff
@@ -196,6 +198,34 @@ def clear_from(self, offset, timestamp):
+        execute = self.cursor.execute
+        if timestamp is not None:
+            params = (translated_key, namespace, timestamp)
+            execute('''DELETE FROM gauged_data WHERE `key` = %s AND namespace = %s AND offset <= %s''', params)
```
The only remaining issue I can see is that you're passing in a `timestamp` here but the table stores block offsets which are equal to `timestamp // config.block_size`.

You'll need to calculate the block offset in the writer (where the config is available) and then pass the `offset` and `timestamp` to the driver methods. Have a look at [`writer#clear_from`](https://github.com/chriso/gauged/blob/37198211702973cbf0aafe08594fc0244d067068/gauged/writer.py#L169-176) and [`driver#clear_from`](https://github.com/chriso/gauged/blob/37198211702973cbf0aafe08594fc0244d067068/gauged/drivers/sqlite.py#L183-190) for an example.

chriso commented:
This is looking good. After the timestamp vs. offset issue is fixed can you squash the commits? I can then merge and publish the new version to pypi. Thanks!

niphlod committed:
fix timestamp and offset parameters

niphlod commented:
ok. ready for squash time ?

niphlod committed:
set timestamp and offset as optional in interface.py

niphlod commented:
okay, done.

niphlod commented:
ok, got the idea. Will do tomorrow. Thanks for the invaluable inputs.

niphlod committed:
clear_key_before should not be inclusive, clear_key_after yes

niphlod commented:
done.

chriso commented on gauged/drivers/interface.py:
```diff
@@ -62,13 +66,19 @@ def remove_namespace(self, namespace):
         pass
 
-    def remove_cache(self, namespace):
+    def remove_cache(self, namespace, key):
```
Should be `def remove_cache(self, namespace, key=None):`

chriso commented on gauged/writer.py:
```diff
@@ -175,6 +175,36 @@ def clear_from(self, timestamp):
+            if remainder:
+                raise ValueError('timestamp must be on a block boundary')
+            if offset > 0:
+                offset -= 1
```
You can't delete before timestamp 0 so it might be worth changing this to:

```python
if offset == 0:
    raise ValueError('cannot delete before offset zero') 
offset -= 1
```

chriso replied:
At the moment if you call `writer.clear_key_before(key, timestamp=0)` it will delete the data between 0 and the first block offset (e.g. `[0, 86399]`) if you're using a daily block size. This is probably not what the user expects.

chriso commented on test/test_gauged.py:
```diff
@@ -549,6 +549,86 @@ def test_clear_from(self):
+        self.assertEqual(gauged.aggregate('foo', Gauged.COUNT, start=50000, end=60000, namespace=1), 0)
+        # 'bar' is still there
+        self.assertEqual(gauged.value('bar', timestamp=60000), 6)
+        self.assertEqual(gauged.value('bar', timestamp=60000, namespace=1), 6)
```
Can you add a test case to both which shows the behavior when you try and delete from a timestamp that isn't on a block boundary?

Something like

```python

with gauged.writer as writer:
    with self.assertRaises(ValueError):
        writer.clear_key_after('foo', timestamp=15000)
```

chriso commented:
Sorry for the back and forth :smile: This is looking great.

niphlod committed:
fixed a few things

niphlod committed:
fixed writer once again

niphlod commented:
should be including all of your revisions, but not this, which I didn't understand...

    At the moment if you call writer.clear_key_before(key, timestamp=0) it will delete the data between 0 and the first block offset (e.g. [0, 86399]) if you're using a daily block size. This is probably not what the user expects.

chriso commented:
> At the moment if you call writer.clear_key_before(key, timestamp=0) it will delete the data between 0 and the first block offset (e.g. [0, 86399]) if you're using a daily block size. This is probably not what the user expects.

I was referring to the [previous discussion](https://github.com/chriso/gauged/pull/5#commitcomment-10154490) where I explained how you can only delete whole blocks at a time.

After changing `clear_key_before` so that it was not inclusive of the specified timestamp, the method was doing the right thing for all timestamps *except for 0*. This was because of the (now fixed) block of code: `if offset > 0: offset -= 1`.

For example, with a block size of 10000, calling `clear_key_before(timestamp=20000)` would correctly delete `block_offset <= 1` (timestamps `[0, 9999]`, and `[10000, 19999]`). If you passed in `clear_key_before(timestamp=0)` however, the `block_offset` would never be decremented and so the driver would delete `block_offset <= 0` (timestamps `[0, 9999]`) which is not what the user would expect.

You fixed it by saying:

```python
if offset == 0:
    raise ValueError('cannot delete before offset zero')
offset -= 1
```

chriso commented:
Thanks again for working on this. gauged==1.0.0 is now available in pypi.
